core/src/test/scala/unit/kafka/server/LogDirFailureTest.scala

```diff
@@ -181,8 +181,21 @@ class LogDirFailureTest extends IntegrationTestHarness {
     // Wait for producer to update metadata for the partition
     TestUtils.waitUntilTrue(() => {
       // ProduceResponse may contain KafkaStorageException and trigger metadata update
-      producer.send(record)
-      producer.partitionsFor(topic).asScala.find(_.partition() == 0).get.leader().id() != leaderServerId
+      {
+        try {
+          producer.send(record).get()
+          true
+        } catch {
+          case e: ExecutionException => {
+            e.getCause match {
+              case t: KafkaStorageException =>
+              case t: NotLeaderForPartitionException => // This may happen if ProduceRequest version <= 3
+              case t: Throwable => fail(s"send() should fail with either KafkaStorageException or NotLeaderForPartitionException instead of ${t.toString}")
+            }
+          }
+            false
+        }
+      } && producer.partitionsFor(topic).asScala.find(_.partition() == 0).get.leader().id() != leaderServerId
```

Comment:
Nit: For readability, could we store the output of the `try` / `catch` block in a variable and use that here? Would also be good to add a comment on why we need both conditions to be true.